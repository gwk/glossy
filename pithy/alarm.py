# Dedicated to the public domain under CC0: https://creativecommons.org/publicdomain/zero/1.0/.

import signal
from typing import Callable


class Timeout(Exception):
  'Exception indictaing that an AlarmManager timed out.'


class AlarmManager:

  def __init__(self, timeout: int, msg:str='alarm timed out after {timeout} seconds', on_signal: Callable[[], None]=None) -> None:
    if not isinstance(timeout, int): raise TypeError(f'timeout must be an int; received: {timeout!r}')
    if timeout < 0: raise TypeError(f'timeout must be nonnegative; received: {timeout!r}')
    self.timeout = timeout
    self.msg = msg
    self.on_signal = on_signal
    self.timed_out = False


  def __enter__(self):
    if not self.timeout: return
    prev = signal.getsignal(signal.SIGALRM)
    if prev is not signal.SIG_DFL:
      raise Exception(f'task.run encountered previously installed signal handler: {prev}')

    def alarm_handler(signum, current_stack_frame):
      # since signal handlers carry reentrancy concerns; do not do any IO within the handler.
      self.timed_out = True
      if self.on_signal: self.on_signal() # callback must respect reentrancy limitations.

    signal.signal(signal.SIGALRM, alarm_handler) # set handler.
    signal.alarm(self.timeout) # set alarm.


  def __exit__(self, exc_type, exc_value, traceback) -> None:
    if not self.timeout: return None
    signal.alarm(0) # disable alarm.
    signal.signal(signal.SIGALRM, signal.SIG_DFL) # uninstall handler.
    if exc_type: return None # exception will be reraised.
    if self.timed_out: raise Timeout(self.msg.format(timeout=self.timeout))
    return None
