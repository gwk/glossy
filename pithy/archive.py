# Dedicated to the public domain under CC0: https://creativecommons.org/publicdomain/zero/1.0/.

from functools import total_ordering
from io import BytesIO, TextIOBase, TextIOWrapper
from typing import Any, BinaryIO, Callable, Iterable, List, Tuple, Union
from tarfile import TarFile
from zipfile import ZipFile, BadZipFile
from .fs import path_ext
from .util import lazy_property
from .io import *


FileOrPath = Union[BinaryIO, str] # TODO: support other pathlike objects.
Opener = Callable[[], BinaryIO]


class Archive:

  def __init__(self, file_or_path:FileOrPath, ext=None) -> None:
    self.name: str = getattr(file_or_path, 'name', str(file_or_path))
    if ext is None:
      ext = path_ext(self.name)
    self.ext = ext
    try: self._handler = _ext_handlers[ext](file_or_path)
    except KeyError: raise ValueError(f'Archive does not support extension: {ext!r}')


  def __repr__(self) -> str:
    return f'{self.__class__.__name__}(name={self.name!r})'


  def __iter__(self) -> Iterable['ArchiveFile']:
    return self._handler.files(self)



@total_ordering
class ArchiveMember:

  def __init__(self, archive: Archive, name: str) -> None:
    self.archive = archive
    self.name = name

  def __repr__(self) -> str:
    return f'{self.__class__.__name__}({self.archive}, {self.name!r})'

  def __lt__(self, other: Any) -> bool:
    if isinstance(other, ArchiveMember):
      return self.name < other.name
    return NotImplemented



class ArchiveFile(ArchiveMember):

  def __init__(self, archive:Archive, name:str, opener:Opener) -> None:
    super().__init__(archive=archive, name=name)
    self._opener = opener

  @lazy_property
  def _file(self): return self._opener()

  def __iter__(self): return iter(self._file)

  def __getattr__(self, name):
    # note: dunder attribute access is never forwarded to this method.
    if name.startswith('_'): raise AttributeError(name)
    return getattr(self._file, name)

  def seekable(self):
    return getattr(self._file, 'seekable', False)

  def text(self, encoding='UTF-8', errors=None, newline=None) -> TextIOWrapper:
    'Returns a TextIOWrapper around the underlying binary buffer.'
    return TextIOWrapper(self._file, encoding=encoding, errors=errors, newline=newline)



class ArchiveDir(ArchiveMember): pass


class _Handler:
  def __init__(self, file_or_path:FileOrPath) -> None: raise NotImplementedError()


class _TarHandler(_Handler):
  def __init__(self, file_or_path:FileOrPath) -> None:
    self.tar = TarFile.open(mode='r|*', fileobj=file_or_path)

  def files(self, archive:Archive) -> Iterable[Tuple[str, Opener]]:
    for info in self.tar:
      if info.isdir(): continue
      opener = lambda: self.tar.extractfile(info)
      #^ note: crucial to pass info here and not name, or else gz archives get very slow.
      yield ArchiveFile(archive=archive, name=info.name, opener=opener)


class _ZipHandler(_Handler):
  def __init__(self, file_or_path:FileOrPath) -> None:
    try: self.zip = ZipFile(file_or_path)
    except BadZipFile as e:
      if isinstance(file_or_path, TextIOBase):
        raise TypeError(f'Archive requires a path or binary stream but received text stream: {file_or_path}') from e
      raise

  def files(self, archive:Archive) -> Iterable[Tuple[str, Opener]]:
    for name in self.zip.namelist():
      if name.endswith('/'): continue # directory.
      opener = lambda: self.zip.open(name)
      yield ArchiveFile(archive=archive, name=name, opener=opener)


_ext_handlers = {
  '.bz2' : _TarHandler,
  '.gz'  : _TarHandler,
  '.tar' : _TarHandler,
  '.xz', : _TarHandler,
  '.zip' : _ZipHandler,
}
