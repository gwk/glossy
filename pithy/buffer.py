# Dedicated to the public domain under CC0: https://creativecommons.org/publicdomain/zero/1.0/.

from typing import Callable, Generic, Iterable, Iterator, List, TypeVar, Union
from .default import Raise, RaiseOr


T = TypeVar('T')


class Buffer(Iterator[T]):
  '''
  Iterable object that buffers an iterable.
  Call push() to push an item into the buffer;
  this will be returned on the subsequent call to next().
  '''


  def __init__(self, iterable: Iterable[T]) -> None:
    self.iterator = iter(iterable)
    self.buffer: List[T] = []


  def __repr__(self) -> str:
    return 'IterBuffer({!r}, buffer={!r})'.format(self.iterator, self.buffer)


  def __iter__(self) -> Iterator[T]: return self


  def __next__(self) -> T:
    try: return self.buffer.pop()
    except IndexError: pass
    return next(self.iterator)


  @property
  def is_live(self) -> bool:
    try: self.peek()
    except StopIteration: return False
    else: return True


  def push(self, item: T) -> None:
    self.buffer.append(item)


  def peek(self, default: RaiseOr[T]=Raise._) -> T:
    try: return self.buffer[-1]
    except IndexError: pass
    try: el = next(self.iterator)
    except StopIteration:
      if isinstance(default, Raise): raise
      else: return default
    self.buffer.append(el)
    return el


  def take_while(self, pred: Callable[[T], bool]) -> Iterator[T]:
    for el in self:
      if pred(el):
        yield el
      else:
        self.buffer.append(el)
        break


  def drop_while(self, pred: Callable[[T], bool]) -> None:
    for el in self:
      if not pred(el):
        self.buffer.append(el)
        break


  def peek_while(self, pred: Callable[[T], bool]) -> List[T]:
    els = list(self.take_while(pred))
    self.buffer.extend(reversed(els))
    return els


  def take(self, count: int, short=False, default: RaiseOr[T]=Raise._) -> List[T]:
    els = []
    for _ in range(count):
      try: els.append(next(self))
      except StopIteration:
        if short: break
        if isinstance(default, Raise): raise
        els.append(default)
    return els


  def peeks(self, count: int, short=False, default: RaiseOr[T]=Raise._) -> List[T]:
    if 0 < count <= len(self.buffer):
      return list(reversed(self.buffer[-count:]))
    els = []
    for _ in range(count):
      try: els.append(next(self))
      except StopIteration:
        if short: break
        if isinstance(default, Raise): raise
        els.append(default)
    self.buffer.extend(reversed(els))
    return els


  def expect(self, pred: Callable[[T], bool]) -> T:
    el = next(self)
    if pred(el): return el
    raise ValueError(el)
