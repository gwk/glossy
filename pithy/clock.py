# Dedicated to the public domain under CC0: https://creativecommons.org/publicdomain/zero/1.0/.

from sys import stderr
from time import time as now


class Clock:
  def __init__(self, name:str):
    self.name = name
    self.times = [(now(), 'start')]

  def split(self, label:str):
    t = now()
    print(f'clock: {self.name}: {label}: {t - self.times[-1][0]:.04f}', file=stderr)
    self.times.append((t, label))


  def total(self):
    t = now()
    print(f'clock: {self.name}: TOTAL: {t - self.times[0][0]:.04f}', file=stderr)
