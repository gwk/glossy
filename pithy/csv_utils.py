# Dedicated to the public domain under CC0: https://creativecommons.org/publicdomain/zero/1.0/.

import csv


def write_csv(f, header, rows):
  w = csv.writer(f)
  w.writerow(header)
  w.writerows(rows)

