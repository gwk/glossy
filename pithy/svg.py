# Dedicated to the public domain under CC0: https://creativecommons.org/publicdomain/zero/1.0/.

'''
SVG writer.
SVG elements reference: https://developer.mozilla.org/en-US/docs/Web/SVG/Element.
'''

from .num import Num, NumRange
from .xml import XmlAttrs, XmlWriter, add_opt_attrs, esc_xml_attr, esc_xml_text
from html import escape as html_escape
from types import TracebackType
from typing import Any, ContextManager, Dict, Iterator, List, Optional, Sequence, TextIO, Tuple, Type, Union, Iterable


Dim = Union[int, float, str]
Vec = Tuple[Num, Num]
VecOrNum = Union[Vec, Num]
PathCommand = Tuple

ViewBox = Union[None, Vec, Tuple[Num, Num, Num, Num], Tuple[Vec, Vec]] # TODO: currently unused.


class SvgWriter(XmlWriter):
  '''
  SvgWriter is a ContextManager class that outputs SVG code to a file (stdout by default).
  Like its parent class XmlWriter, it uses the __enter__ and __exit__ methods to automatically output open and close tags.
  '''

  replaced_attrs = {
    'href': 'xlink:href', # safari Version 11.1.1 requires this, even though xlink is deprecated in svg 2 standard.
    **XmlWriter.replaced_attrs,
  }


  def __init__(self, file:TextIO=None, x:Dim=None, y:Dim=None, w:Dim=None, h:Dim=None,
   vx:Num=None, vy:Num=None, vw:Num=None, vh:Num=None, **attrs:Any) -> None:
    self.x = x
    self.y = y
    self.w = w
    self.h = h
    self.vx = vx
    self.vy = vy
    self.vw = vw
    self.vh = vh
    self.viewBox = fmt_viewBox(vx, vy, vw, vh)
    attrs = { # Put the xml nonsense up front.
      'xmlns': 'http://www.w3.org/2000/svg',
      'xmlns:xlink': 'http://www.w3.org/1999/xlink', # Safari does not yet support SVG 2.
      **attrs
    }
    add_opt_attrs(attrs, x=x, y=y, width=w, height=h, viewBox=self.viewBox)
    super().__init__(tag='svg', file=file, attrs=attrs)


  def leaf(self, tag:str, attrs:XmlAttrs) -> None:
    '''
    Output a non-nesting SVG element.
    'title' is treated as a special attribute that is translated into a <title> child element and rendered as a tooltip.
    '''
    if attrs and 'title' in attrs: self.leaf_text(tag, attrs, text='')
    else: super().leaf(tag, attrs)


  def leaf_text(self, tag:str, attrs:XmlAttrs, text:str) -> None:
    '''
    Output a non-nesting XML element that contains text between the open and close tags.
    'title' is treated as a special attribute that is translated into a <title> child element and rendered as a tooltip.
    '''
    if attrs:
      try: title = attrs.pop('title')
      except KeyError: pass
      else:
        self.write(f'<{tag}{self.fmt_attrs(attrs)}>{self.title(title)}{esc_xml_text(text)}</{tag}>')
        return
    super().leaf_text(tag, attrs=attrs, text=text)


  def sub(self, tag:str, attrs:XmlAttrs) -> XmlWriter:
    '''
    Create a child XmlWriter for use in a `with` context to represent a nesting XML element.
    'title' is treated as a special attribute that is translated into a <title> child element and rendered as a tooltip.
    '''
    title_el = ''
    if attrs:
      try: title = attrs.pop('title')
      except KeyError: pass
      else:
        title_el = self.title(title)
    s = super().sub(tag, attrs=attrs)
    if title_el: self.write(title_el)
    return s


  # SVG Elements.

  def title(self, title:Optional[str]) -> str:
    return '' if title is None else f'<title>{esc_xml_text(title)}</title>'


  def circle(self, pos:Vec=None, r:Num=None, *, x:Num=None, y:Num=None, **attrs) -> None:
    'Output an SVG `circle` element.'
    if pos is not None:
      assert x is None
      assert y is None
      x, y = pos
    add_opt_attrs(attrs, cx=x, cy=y, r=r)
    self.leaf('circle', attrs)


  def defs(self, **attrs:Any) -> XmlWriter:
    'Output an SVG `defs` element.'
    return self.sub('defs', attrs)


  def g(self, *transforms:str, **attrs:Any) -> XmlWriter:
    'Create an SVG `g` element for use in a context manager.'
    add_opt_attrs(attrs, transform=(' '.join(transforms) if transforms else None))
    return self.sub('g', attrs)


  def image(self, pos:Vec=None, size:VecOrNum=None, *, x:Num=None, y:Num=None, w:Num=None, h:Num=None, **attrs) -> None:
    'Output an SVG `defs` element.'
    if pos is not None:
      assert x is None
      assert y is None
      x, y = pos
    if size is not None:
      assert w is None
      assert h is None
      if isinstance(size, tuple):
        w, h = size
      else:
        w = h = size
    add_opt_attrs(attrs, x=x, y=y, width=w, height=h)
    self.leaf('image', attrs)


  def line(self, a:Vec=None, b:Vec=None, *, x1:Num=None, y1:Num=None, x2:Num=None, y2:Num=None, **attrs) -> None:
    'Output an SVG `defs` element.'
    if a is not None:
      assert x1 is None
      assert y1 is None
      x1, y1 = a
    if b is not None:
      assert x2 is None
      assert y2 is None
      x2, y2 = b
    add_opt_attrs(attrs, x1=x1, y1=y1, x2=x2, y2=y2)
    self.leaf('line', attrs)


  def marker(self, id:str, pos:Vec=None, size:VecOrNum=None, *, x:Num=None, y:Num=None, w:Num=None, h:Num=None,
   vx:Num=None, vy:Num=None, vw:Num=None, vh:Num=None,
   markerUnits='strokeWidth', orient:str='auto', **attrs:Any) -> XmlWriter:
    'Output an SVG `marker` element.'
    if pos is not None:
      assert x is None
      assert y is None
      x, y = pos
    if size is not None:
      assert w is None
      assert h is None
      if isinstance(size, tuple):
        w, h = size
      else:
        w = h = size
    add_opt_attrs(attrs, id=id, refX=x, refY=y, markerWidth=w, markerHeight=h, viewBox=fmt_viewBox(vx, vy, vw, vh),
      markerUnits=markerUnits, orient=orient)
    return self.sub('marker', attrs=attrs)


  def path(self, commands:Iterable[PathCommand], **attrs:Any) -> None:
    'Output an SVG `path` element.'
    assert 'd' not in attrs
    cmd_strs:List[str] = []
    for c in commands:
      try: code = c[0]
      except IndexError: continue
      try: exp_len = path_command_lens[code]
      except KeyError as e: raise Exception(f'bad path command code: {c!r}') from e
      if len(c) != exp_len + 1: raise Exception(f'path command requires {exp_len} arguments: {c}')
      cmd_strs.append(code + ','.join(str(x) for x in c[1:]))
    assert 'd' not in attrs
    attrs['d'] = ' '.join(cmd_strs)
    self.leaf('path', attrs)


  def polyline(self, points:Iterable[Vec], **attrs:Any) -> None:
    point_strs:List[str] = []
    assert 'points' not in attrs
    for p in points:
      if len(p) < 2: raise Exception(f'bad point for polyline: {p}')
      point_strs.append(f'{p[0]},{p[1]}')
    attrs['points'] = ' '.join(point_strs)
    self.leaf('polyline', attrs)


  def rect(self, pos:Vec=None, size:VecOrNum=None, *, x:Num=None, y:Num=None, w:Num=None, h:Num=None, r:VecOrNum=None, **attrs:Any) -> None:
    'Output an SVG `rect` element.'
    if pos is not None:
      assert x is None
      assert y is None
      x, y = pos
    if size is not None:
      assert w is None
      assert h is None
      if isinstance(size, tuple):
        w, h = size
      else:
        w = h = size
    rx:Optional[Num]
    ry:Optional[Num]
    if isinstance(r, tuple):
      rx, ry = r
    else:
      rx = ry = r
    add_opt_attrs(attrs, x=x, y=y, width=w, height=h, rx=rx, ry=ry)
    self.leaf('rect', attrs)


  def style(self, text:str, **attrs,) -> None:
    'Output an SVG `style` element.'
    self.leaf_text('style', attrs, text.strip())


  def symbol(self, id:str, vx:Num=None, vy:Num=None, vw:Num=None, vh:Num=None, **attrs:Any) -> XmlWriter:
    'Output an SVG `symbol` element.'
    if vx is None: vx = 0
    if vy is None: vy = 0
    # TODO: figure out if no viewBox is legal and at all useful.
    assert vw >= 0 # type: ignore
    assert vh >= 0 # type: ignore
    add_opt_attrs(attrs, id=id, viewBox=f'{vx} {vy} {vw} {vh}')
    return self.sub('symbol', attrs)


  def text(self, pos:Vec=None, x:Num=None, y:Num=None, text=None, **attrs) -> None:
    'Output an SVG `text` element.'
    if pos is not None:
      assert x is None
      assert y is None
      x, y = pos
    add_opt_attrs(attrs, x=x, y=y)
    self.leaf_text('text', attrs, text)


  def use(self, id:str, pos:Vec=None, size:VecOrNum=None, x:Num=None, y:Num=None, w:Num=None, h:Num=None, **attrs) -> None:
    'Use a previously defined symbol'
    assert id
    if id[0] != '#': id = '#' + id
    if pos is not None:
      assert x is None
      assert y is None
      x, y = pos
    if size is not None:
      assert w is None
      assert h is None
      if isinstance(size, tuple):
        w, h = size
      else:
        w = h = size
    add_opt_attrs(attrs, href=id, x=x, y=y, width=w, height=h)
    return self.leaf('use', attrs)


  # High level.

  def grid(self, pos:Vec=None, size:VecOrNum=None, *,
   x:Num=0, y:Num=0, w:Num=256, h:Num=256, step:VecOrNum=16, off:Vec=(0, 0), r:VecOrNum=None, **attrs:Any) -> None:
    # TODO: axis transformers (e.g. log-log).
    assert step is not None
    if isinstance(step, tuple):
      sx, sy = step
    else:
      sx = sy = step
    if pos is not None:
      x, y = pos
    if size is not None:
      if isinstance(size, tuple):
        w, h = size
      else:
        w = h = size
    x = float(x)
    y = float(y)
    w = float(w)
    h = float(h)
    off_x, off_y = off
    if off_x <= 0: off_x = sx
    if off_y <= 0: off_y = sy
    x_start = x + off_x
    y_start = y + off_y
    x_end = x + w
    y_end = y + h
    class_ = attrs.pop('class_', 'grid')
    with self.g(class_=class_, **attrs):
      self.rect(class_=class_+'-border', x=x, y=y, w=w, h=h, r=r)
      for tick in NumRange(x_start, x_end, sx): self.line((tick, y), (tick, y_end)) # Vertical lines.
      for tick in NumRange(y_start, y_end, sy): self.line((x, tick), (x_end, tick)) # Horizontal lines.


def scale(x:Num=1, y:Num=1) -> str: return f'scale({x},{y})'
def rotate(degrees:Num, x:Num=0, y:Num=0) -> str: return f'rotate({degrees},{x},{y})'
def translate(x:Num=0, y:Num=0) -> str: return f'translate({x},{y})'


path_command_lens = {
  'A' : 7,
  'M' : 2,
  'm' : 2,
  'L' : 2,
  'l' : 2,
  'Z' : 0,
  'z' : 0
}

valid_units = frozenset({'em', 'ex', 'px', 'pt', 'pc', 'cm', 'mm', '%', ''})

def _validate_unit(unit: str):
  'Ensure that `unit` is a valid unit string.'
  if unit not in valid_units:
    raise Exception(f'Invalid SVG unit: {unit!r}; should be one of {sorted(valid_units)}')


def fmt_viewBox(vx:Optional[Num], vy:Optional[Num], vw:Optional[Num], vh:Optional[Num]) -> Optional[str]:
  if vx is None and vy is None and vw is None and vh is None:
    return None
  else:
    if vx is None: vx = 0
    if vy is None: vy = 0
    assert vw is not None and vw > 0
    assert vh is not None and vh > 0
    return f'{vx} {vy} {vw} {vh}'


baselines = {
  'after-edge',
  'alphabetic',
  'auto',
  'baseline',
  'before-edge',
  'central',
  'hanging',
  'ideographic',
  'inherit',
  'mathematical',
  'middle',
  'text-after-edge',
  'text-before-edge',
}
