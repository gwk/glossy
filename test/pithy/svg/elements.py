#!/usr/bin/env python3

from pithy.svg import *
from sys import stdout

w = 512
h = 1024
with Svg(w=w, h=h) as svg:
  svg.grid(size=(w, h), stroke='#E0E0E0', fill='white', stroke_width=0.5)

  y = 0

  with svg.g(translate(y=y)) as g:
    g.rect((0, 0), (64, 32), title='rect0', r=8)
    g.rect(x=80, y=0, w=64, h=32, r=(16, 8), stroke='gray', fill='white', title='rect1')
  y += 48

  with svg.g(translate(y=y)) as g:
    g.circle((16, 16), 16, title='circle0')
    g.circle(x=64, y=16, r=16, stroke='gray', fill='white', title='circle1')
  y += 48

  with svg.g(translate(y=y)) as g:
    g.line((0, 0), (32, 32), stroke='black', title='line0')
    g.line(x1=32, y1=0, x2=64, y2=32, stroke='black', title='line1')
  y += 48

  with svg.g(translate(y=y)) as g:
    g.path([('M', 0, 0), ('L', 32, 32), ('l', 32, -32)], stroke='black', fill=None, title='path0')
  y += 48

  with svg.g(translate(y=y)) as g:
    g.polyline([(0,0), (32,32), (64,0)], stroke='black', fill=None, title='polyline0')
  y += 48

  svg.text(x=0, y=y, text='TEXT')
  y += 16

  img_base64 = 'data:image/png;base64,R0lGODdhBAAEAIAAAAAAAP///yH5BAQAAAAALAAAAAAEAAQAAAIFRHxnuAUAOw=='

  svg.image((0, y), (16,16), href=img_base64)
  svg.image(x=32, y=y, w=32, h=32, href=img_base64)
  y += 48

  with svg.g(translate(y=y)) as g:
    with g.symbol(id='sym', vw=32, vh=32) as g1:
      g1.rect(pos=(8, 8), size=(16, 16), stroke='black', fill=None)
      g1.rect(pos=(0, 0), size=(32, 32), stroke='black', fill=None)
    g.use('sym', (0,0), 32)
    g.use('sym', (48,0), 16)
  y += 48

  with svg.g(translate(x=16, y=y)) as g:
    with g.marker(id='dot', pos=(4,4), size=8, vw=8, vh=8) as marker:
      marker.circle((4, 4), r=4)
    g.polyline([(0,0), (16,0), (32,0)], stroke='black', marker_start="url(#dot)", marker_mid="url(#dot)", marker_end="url(#dot)")
  y += 32

svg.write(stdout)
