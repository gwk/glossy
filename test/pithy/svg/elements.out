<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="512" height="1024">
<g class="grid" stroke="#E0E0E0" fill="white" stroke-width="0.5">
<rect class="grid-border" x="0.0" y="0.0" width="512.0" height="1024.0"/>
<line x1="16.0" y1="0.0" x2="16.0" y2="1024.0"/>
<line x1="32.0" y1="0.0" x2="32.0" y2="1024.0"/>
<line x1="48.0" y1="0.0" x2="48.0" y2="1024.0"/>
<line x1="64.0" y1="0.0" x2="64.0" y2="1024.0"/>
<line x1="80.0" y1="0.0" x2="80.0" y2="1024.0"/>
<line x1="96.0" y1="0.0" x2="96.0" y2="1024.0"/>
<line x1="112.0" y1="0.0" x2="112.0" y2="1024.0"/>
<line x1="128.0" y1="0.0" x2="128.0" y2="1024.0"/>
<line x1="144.0" y1="0.0" x2="144.0" y2="1024.0"/>
<line x1="160.0" y1="0.0" x2="160.0" y2="1024.0"/>
<line x1="176.0" y1="0.0" x2="176.0" y2="1024.0"/>
<line x1="192.0" y1="0.0" x2="192.0" y2="1024.0"/>
<line x1="208.0" y1="0.0" x2="208.0" y2="1024.0"/>
<line x1="224.0" y1="0.0" x2="224.0" y2="1024.0"/>
<line x1="240.0" y1="0.0" x2="240.0" y2="1024.0"/>
<line x1="256.0" y1="0.0" x2="256.0" y2="1024.0"/>
<line x1="272.0" y1="0.0" x2="272.0" y2="1024.0"/>
<line x1="288.0" y1="0.0" x2="288.0" y2="1024.0"/>
<line x1="304.0" y1="0.0" x2="304.0" y2="1024.0"/>
<line x1="320.0" y1="0.0" x2="320.0" y2="1024.0"/>
<line x1="336.0" y1="0.0" x2="336.0" y2="1024.0"/>
<line x1="352.0" y1="0.0" x2="352.0" y2="1024.0"/>
<line x1="368.0" y1="0.0" x2="368.0" y2="1024.0"/>
<line x1="384.0" y1="0.0" x2="384.0" y2="1024.0"/>
<line x1="400.0" y1="0.0" x2="400.0" y2="1024.0"/>
<line x1="416.0" y1="0.0" x2="416.0" y2="1024.0"/>
<line x1="432.0" y1="0.0" x2="432.0" y2="1024.0"/>
<line x1="448.0" y1="0.0" x2="448.0" y2="1024.0"/>
<line x1="464.0" y1="0.0" x2="464.0" y2="1024.0"/>
<line x1="480.0" y1="0.0" x2="480.0" y2="1024.0"/>
<line x1="496.0" y1="0.0" x2="496.0" y2="1024.0"/>
<line x1="0.0" y1="16.0" x2="512.0" y2="16.0"/>
<line x1="0.0" y1="32.0" x2="512.0" y2="32.0"/>
<line x1="0.0" y1="48.0" x2="512.0" y2="48.0"/>
<line x1="0.0" y1="64.0" x2="512.0" y2="64.0"/>
<line x1="0.0" y1="80.0" x2="512.0" y2="80.0"/>
<line x1="0.0" y1="96.0" x2="512.0" y2="96.0"/>
<line x1="0.0" y1="112.0" x2="512.0" y2="112.0"/>
<line x1="0.0" y1="128.0" x2="512.0" y2="128.0"/>
<line x1="0.0" y1="144.0" x2="512.0" y2="144.0"/>
<line x1="0.0" y1="160.0" x2="512.0" y2="160.0"/>
<line x1="0.0" y1="176.0" x2="512.0" y2="176.0"/>
<line x1="0.0" y1="192.0" x2="512.0" y2="192.0"/>
<line x1="0.0" y1="208.0" x2="512.0" y2="208.0"/>
<line x1="0.0" y1="224.0" x2="512.0" y2="224.0"/>
<line x1="0.0" y1="240.0" x2="512.0" y2="240.0"/>
<line x1="0.0" y1="256.0" x2="512.0" y2="256.0"/>
<line x1="0.0" y1="272.0" x2="512.0" y2="272.0"/>
<line x1="0.0" y1="288.0" x2="512.0" y2="288.0"/>
<line x1="0.0" y1="304.0" x2="512.0" y2="304.0"/>
<line x1="0.0" y1="320.0" x2="512.0" y2="320.0"/>
<line x1="0.0" y1="336.0" x2="512.0" y2="336.0"/>
<line x1="0.0" y1="352.0" x2="512.0" y2="352.0"/>
<line x1="0.0" y1="368.0" x2="512.0" y2="368.0"/>
<line x1="0.0" y1="384.0" x2="512.0" y2="384.0"/>
<line x1="0.0" y1="400.0" x2="512.0" y2="400.0"/>
<line x1="0.0" y1="416.0" x2="512.0" y2="416.0"/>
<line x1="0.0" y1="432.0" x2="512.0" y2="432.0"/>
<line x1="0.0" y1="448.0" x2="512.0" y2="448.0"/>
<line x1="0.0" y1="464.0" x2="512.0" y2="464.0"/>
<line x1="0.0" y1="480.0" x2="512.0" y2="480.0"/>
<line x1="0.0" y1="496.0" x2="512.0" y2="496.0"/>
<line x1="0.0" y1="512.0" x2="512.0" y2="512.0"/>
<line x1="0.0" y1="528.0" x2="512.0" y2="528.0"/>
<line x1="0.0" y1="544.0" x2="512.0" y2="544.0"/>
<line x1="0.0" y1="560.0" x2="512.0" y2="560.0"/>
<line x1="0.0" y1="576.0" x2="512.0" y2="576.0"/>
<line x1="0.0" y1="592.0" x2="512.0" y2="592.0"/>
<line x1="0.0" y1="608.0" x2="512.0" y2="608.0"/>
<line x1="0.0" y1="624.0" x2="512.0" y2="624.0"/>
<line x1="0.0" y1="640.0" x2="512.0" y2="640.0"/>
<line x1="0.0" y1="656.0" x2="512.0" y2="656.0"/>
<line x1="0.0" y1="672.0" x2="512.0" y2="672.0"/>
<line x1="0.0" y1="688.0" x2="512.0" y2="688.0"/>
<line x1="0.0" y1="704.0" x2="512.0" y2="704.0"/>
<line x1="0.0" y1="720.0" x2="512.0" y2="720.0"/>
<line x1="0.0" y1="736.0" x2="512.0" y2="736.0"/>
<line x1="0.0" y1="752.0" x2="512.0" y2="752.0"/>
<line x1="0.0" y1="768.0" x2="512.0" y2="768.0"/>
<line x1="0.0" y1="784.0" x2="512.0" y2="784.0"/>
<line x1="0.0" y1="800.0" x2="512.0" y2="800.0"/>
<line x1="0.0" y1="816.0" x2="512.0" y2="816.0"/>
<line x1="0.0" y1="832.0" x2="512.0" y2="832.0"/>
<line x1="0.0" y1="848.0" x2="512.0" y2="848.0"/>
<line x1="0.0" y1="864.0" x2="512.0" y2="864.0"/>
<line x1="0.0" y1="880.0" x2="512.0" y2="880.0"/>
<line x1="0.0" y1="896.0" x2="512.0" y2="896.0"/>
<line x1="0.0" y1="912.0" x2="512.0" y2="912.0"/>
<line x1="0.0" y1="928.0" x2="512.0" y2="928.0"/>
<line x1="0.0" y1="944.0" x2="512.0" y2="944.0"/>
<line x1="0.0" y1="960.0" x2="512.0" y2="960.0"/>
<line x1="0.0" y1="976.0" x2="512.0" y2="976.0"/>
<line x1="0.0" y1="992.0" x2="512.0" y2="992.0"/>
<line x1="0.0" y1="1008.0" x2="512.0" y2="1008.0"/>
</g>
<g transform="translate(0,0)">
<rect x="0" y="0" width="64" height="32" rx="8" ry="8"><title>rect0</title></rect>
<rect stroke="gray" fill="white" x="80" y="0" width="64" height="32" rx="16" ry="8"><title>rect1</title></rect>
</g>
<g transform="translate(0,48)">
<circle cx="16" cy="16" r="16"><title>circle0</title></circle>
<circle stroke="gray" fill="white" cx="64" cy="16" r="16"><title>circle1</title></circle>
</g>
<g transform="translate(0,96)">
<line stroke="black" x1="0" y1="0" x2="32" y2="32"><title>line0</title></line>
<line stroke="black" x1="32" y1="0" x2="64" y2="32"><title>line1</title></line>
</g>
<g transform="translate(0,144)">
<path stroke="black" fill="none" d="M0,0 L32,32 l32,-32"><title>path0</title></path>
</g>
<g transform="translate(0,192)">
<polyline stroke="black" fill="none" points="0,0 32,32 64,0"><title>polyline0</title></polyline>
</g>
<text x="0" y="240">TEXT</text>
<image xlink:href="data:image/png;base64,R0lGODdhBAAEAIAAAAAAAP///yH5BAQAAAAALAAAAAAEAAQAAAIFRHxnuAUAOw==" x="0" y="256" width="16" height="16"/>
<image xlink:href="data:image/png;base64,R0lGODdhBAAEAIAAAAAAAP///yH5BAQAAAAALAAAAAAEAAQAAAIFRHxnuAUAOw==" x="32" y="256" width="32" height="32"/>
<g transform="translate(0,304)">
<symbol id="sym" viewBox="0 0 32 32">
<rect stroke="black" fill="none" x="8" y="8" width="16" height="16"/>
<rect stroke="black" fill="none" x="0" y="0" width="32" height="32"/>
</symbol>
<use xlink:href="#sym" x="0" y="0" width="32" height="32"/>
<use xlink:href="#sym" x="48" y="0" width="16" height="16"/>
</g>
<g transform="translate(16,352)">
<marker id="dot" refX="4" refY="4" markerWidth="8" markerHeight="8" viewBox="0 0 8 8" markerUnits="strokeWidth" orient="auto">
<circle cx="4" cy="4" r="4"/>
</marker>
<polyline stroke="black" marker-start="url(#dot)" marker-mid="url(#dot)" marker-end="url(#dot)" points="0,0 16,0 32,0"/>
</g>
</svg>
